doc/crm.8.adoc

```diff
@@ -2104,6 +2104,71 @@ Example:
 utilization xen1 set memory 4096
 ...............
 
+[[cmdhelp.sbd,SBD management]]
+=== `sbd` - SBD management
+
+This level is for managing the SBD (STONITH Block Device) daemon.
+
+[[cmdhelp.sbd.configure,configure SBD]]
+==== `configure`
+
+Configure the SBD daemon for both disk-based and disk-less mode.
+
+Main functionailities include:
+- Show configured disk metadata
+- Show contents of /etc/sysconfig/sbd
+- Show SBD related cluster properties
+- Newly setup SBD configuration on a running cluster
+- Update the existing parameters
+- Add more devices to the existing disk-based SBD configuration
+
+For more details on SBD and related parameters, please see man sbd(8).
+
+Usage:
+...............
+# For disk-based SBD
+crm sbd configure show [disk_metadata|sysconfig|property]
+crm sbd configure [device=<dev>]... [watchdog-device=<dev>] [watchdog-timeout=<integer>] [allocate-timeout=<integer>] [loop-timeout=<integer>] [msgwait-timeout=<integer>]
```

Comment:
BUG: need exit immediately when choose "n" to reject overwriting sbd disk, otherwise it can create the inconsistent metadata, which is not supported.

```
tws-corosync3-2:~ # crm sbd configure device=/dev/vdb29
/dev/vdb29 has already been initialized by SBD, do you want to overwrite it (y/n)? n
INFO: Update SBD_DEVICE in /etc/sysconfig/sbd: /dev/vdb28;/dev/vdb29

tws-corosync3-2:~ # crm sbd configure show disk_metadata
INFO: crm sbd configure show disk_metadata
==Dumping header on disk /dev/vdb28
Header version     : 2.1
UUID               : 192cc144-37f5-4441-9d59-d547a30fcfae
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 2
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 10
==Header on disk /dev/vdb28 is dumped

==Dumping header on disk /dev/vdb29
Header version     : 2.1
UUID               : a4f362d9-454f-4221-9053-b344f9a1dcde
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 20
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/vdb29 is dumped

```

Reply:
> BUG: need exit immediately when choose "n" to reject overwriting sbd disk, otherwise it can create the inconsistent metadata, which is not supported.

This is our current logic, see the process to configure sbd in bootstrap stage, interactive mode
```
15sp6-1:~ # crm cluster init sbd
INFO: Loading "default" profile from /etc/crm/profiles.yml
WARNING: chronyd.service is not configured to start at system boot.
Do you want to continue anyway (y/n)? y
/etc/pacemaker/authkey already exists - overwrite (y/n)? y
INFO: Configure SBD:
  If you have shared storage, for example a SAN or iSCSI target,
  you can use it avoid split-brain scenarios by configuring SBD.
  This requires a 1 MB partition, accessible to all nodes in the
  cluster.  The device path must be persistent and consistent
  across all nodes in the cluster, so /dev/disk/by-id/* devices
  are a good choice.  Note that all data on the partition you
  specify here will be destroyed.

Do you wish to use SBD (y/n)? y
Path to storage device (e.g. /dev/disk/by-id/...), or "none" for diskless sbd, use ";" as separator for multi path []/dev/sda5;/dev/sda6
SBD is already configured to use /dev/sda5 - overwrite (y/n)? n
SBD is already configured to use /dev/sda6 - overwrite (y/n)? n
INFO: Restarting cluster service
INFO: BEGIN Waiting for cluster
...........                                                                                                                                                                                        INFO: END Waiting for cluster
WARNING: "stonith-enabled" in crm_config is set to true, it was false
INFO: BEGIN Adjusting sbd related timeout values
WARNING: "stonith-timeout" in crm_config is set to 71, it was 43
INFO: END Adjusting sbd related timeout values
INFO: Done (log saved to /var/log/crmsh/crmsh.log)
15sp6-1:~ # sbd -d /dev/sda5 dump
==Dumping header on disk /dev/sda5
Header version     : 2.1
UUID               : 5d38e880-e644-4648-93b1-788f744d5559
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 15
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 30
==Header on disk /dev/sda5 is dumped
15sp6-1:~ # sbd -d /dev/sda6 dump
==Dumping header on disk /dev/sda6
Header version     : 2.1
UUID               : 96029eac-a917-430a-8670-9661cdadc304
Number of slots    : 255
Sector size        : 512
Timeout (watchdog) : 5
Timeout (allocate) : 2
Timeout (loop)     : 1
Timeout (msgwait)  : 10
==Header on disk /dev/sda6 is dumped
15sp6-1:~ # cat /etc/sysconfig/sbd |grep DEVICE
# SBD_DEVICE specifies the devices to use for exchanging sbd messages
#SBD_DEVICE=""
SBD_DEVICE="/dev/sda5;/dev/sda6"
```

Also need to change here?

Reply:
In this process, sbd healthcheck logic should be useful and important. It must return the critical error in this situation in my opinion, and tell the user how to fix it. What would be the idea then?

crm cluster health sbd --fix
crm sbd device copy dev1 dev2

any other idea?

Reply:
> In this process, sbd healthcheck logic should be useful and important. It must return the critical error in this situation in my opinion, and tell the user how to fix it. What would be the idea then?
> 
> crm cluster health sbd --fix crm sbd device copy dev1 dev2
> 
Are we going to add `crm sbd device` interface this time?

> any other idea?

Reply:
FYI, I don't have other elegant approach to copy dev1 to dev2, except "dd if=dev1 of=dev2 bs=XXX count=1". XXX usually is 512, but can be different, for example, in s390 as said in the man page.

Reply:
I think if users reject to overwrite the metadata, initialization process should abort immediately. If `crm cluster init sbd` does not do that in the past, it should be changed.

Think about in which case a user would reject to overwrite the metadata. Maybe a wrong device is specified. It does not make sense to continue setting up.

Reply:
> FYI, I don't have other elegant approach to copy dev1 to dev2, except "dd if=dev1 of=dev2 bs=XXX count=1". XXX usually is 512, but can be different, for example, in s390 as said in the man page.

Maybe we can read it from `/sys/class/block/<device>/queue/logical_block_size`?

Reply:
> Think about in which case a user would reject to overwrite the metadata. Maybe a wrong device is specified.
I remember I heard the corner case that a user want to keep the disk metadata unchanged and reuse it.

Reply:
>> FYI, I don't have other elegant approach to copy dev1 to dev2, except "dd if=dev1 of=dev2 bs=XXX count=1". XXX usually is 512, but can be different, for example, in s390 as said in the man page.
> Maybe we can read it from /sys/class/block/<device>/queue/logical_block_size?

The better interface is `strace blockdev --getss /dev/vdb28 2>&1 | grep ioctl`, for example. It calls ` ioctl(3, BLKSSZGET, [512])` and matches to the SBD code [1]


[1]: https://github.com/ClusterLabs/sbd/blob/cf5c2208bad2db2dff9b09624b89b05415c3bc11/src/sbd-md.c#L135

Reply:
> > Think about in which case a user would reject to overwrite the metadata. Maybe a wrong device is specified.
> 
> I remember I heard the corner case that a user want to keep the disk metadata unchanged and reuse it.

https://bugzilla.suse.com/show_bug.cgi?id=1201428